### PR TITLE
Refactor `CreateHardLink` PInvoke

### PR DESCRIPTION
**Pull request by xtqqczze**
WIP

### CODE REVIEW

**Comment by daxian-dbw** [comment]
@xtqqczze Isn't `nint` maps to `IntPtr`, just like `int` maps to `Int32`?

**Comment by xtqqczze** [comment]
It is true that in .NET 7, `nint` is an alias for `System.IntPtr`, however by convention PInvoke signatures use `IntPtr` to represent opaque pointers.

**Comment by xtqqczze** [comment]
error SYSLIB1051: The type 'IntPtr' is not supported by source-generated P/Invokes. The generated source will not handle marshalling of parameter 'lpSecurityAttributes'. [D:\a\1\s\src\System.Management.Automation\System.Management.Automation.csproj]

**Comment by daxian-dbw** [comment]
@xtqqczze The CI of the PR that introduced this file passed: https://github.com/PowerShell/PowerShell/pull/18603. Does that mean we indeed should use `nint` instead of `IntPtr`?

**Comment by xtqqczze** [comment]
> @xtqqczze The CI of the PR that introduced this file passed: #18603. Does that mean we indeed should use `nint` instead of `IntPtr`?

@daxian-dbw I guess that's why @iSazonov replaced `IntPtr` with `nint` in that PR.

I think the best option here is to use `void*` and wrap the PInvoke as in 6e52da5. The _correct_ type is `LPSECURITY_ATTRIBUTES*`, but documentation specifies that `lpSecurityAttributes` is reserved and must be null.

**Comment by iSazonov** [comment]
The old style (using IntPtr and void*) forces using _unsafe_ context.
Now .Net team introduced `nint` to use _pointer arithmetic_ and to avoid unsafe.
We will have analyzer/fixer for recommended patterns https://github.com/dotnet/runtime/issues/74518
For better understanding see also:
https://github.com/dotnet/csharplang/issues/6065
https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/numeric-intptr.md

File: src/System.Management.Automation/engine/Interop/Windows/CreateHardLink.cs [hunk 1]
```diff
@@ -10,7 +10,14 @@ internal static partial class Interop
     internal static unsafe partial class Windows
     {
         [LibraryImport("api-ms-win-core-file-l2-1-0.dll", EntryPoint = "CreateHardLinkW", StringMarshalling = StringMarshalling.Utf16)]
-        [return: MarshalAs(UnmanagedType.Bool)]
-        internal static partial bool CreateHardLink(string name, string existingFileName, nint securityAttributes);
+        private static unsafe partial BOOL CreateHardLink(string lpFileName, string lpExistingFileName, void* lpSecurityAttributes);
```
Comment (jborean93) [inline src/System.Management.Automation/engine/Interop/Windows/CreateHardLink.cs:13]: Is there any benefit to using a custom `BOOL` enum here? Dotnet can handle all that for you as it was before. Why use `void*` when `IntPtr` would suffice and remove the need for the unsafe context.

**Comment by SeeminglyScience** [comment]
> The old style (using IntPtr and void*) forces using _unsafe_ context. Now .Net team introduced `nint` to use _pointer arithmetic_ and to avoid unsafe. We will have analyzer/fixer for recommended patterns [dotnet/runtime#74518](https://github.com/dotnet/runtime/issues/74518) For better understanding see also: [dotnet/csharplang#6065](https://github.com/dotnet/csharplang/issues/6065) [dotnet/csharplang@`main`/proposals/csharp-11.0/numeric-intptr.md](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/numeric-intptr.md?rgh-link-date=2022-12-02T04%3A37%3A59Z)

`nint` was introduced to allow using native sized integers as you would other numbers. Primarily for when it *does not* represent an actual pointer. Despite the name, there's nothing inherently more unsafe about `unsafe` contexts compared to using `IntPtr`.

Now while I did want to clarify that point... if we're never going to actually use the parameter then I agree that `nint` works just fine.

**Comment by daxian-dbw** [comment]
I'm really confused 😅 Can someone please explain the following questions to me?
1. Is `nint` interchangeable with `IntPtr` when used as a parameter type like in this case?
2. Why this PR is needed? (why is using `nint` here is not appropriate?)

**Comment by iSazonov** [comment]
> I'm really confused 😅 Can someone please explain the following questions to me?
> 
> 1. Is `nint` interchangeable with `IntPtr` when used as a parameter type like in this case?

Yes. From design docs referenced above
>In short, we now treat nint/nuint as simple types aliasing System.IntPtr/System.UIntPtr, like we do for int in relation to System.Int32. 

> 2. Why this PR is needed? (why is using `nint` here is not appropriate?)

We have no need the PR. The intention of the .Net team is to explicitly use nint where arithmetic is used and most likely in other cases (although there are doubts about readability)
Also from CreateHardLink docs:
> lpSecurityAttributes
Reserved; must be NULL.

**Comment by daxian-dbw** [comment]
Thanks @iSazonov. In this case then, @xtqqczze shall we close this PR?

**Comment by xtqqczze** [comment]
> The old style (using IntPtr and void*) forces using _unsafe_ context. Now .Net team introduced `nint` to use _pointer 

There is no reason to avoid unsafe contexts in Pinvokes. In fact, it is an error ([SYSLIB1062](https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1050-1069)) to use [LibraryImportAttribute](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.libraryimportattribute) without AllowUnsafeBlocks set to true.

File: src/System.Management.Automation/engine/Interop/Windows/CreateHardLink.cs (hunk 1 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/Interop/Windows/CreateHardLink.cs]: I've removed the enum and explained (https://github.com/PowerShell/PowerShell/pull/18707#issuecomment-1335819527) why we shouldn't avoid unsafe context.

**Comment by xtqqczze** [comment]
> We have no need the PR. The intention of the .Net team is to explicitly use nint where pointer arithmetic is used and most likely in other cases (although there are doubts about readability)

I believe the .NET team recommends replacing IntPtr with actual pointer types if the parameter is a pointer, e.g. https://github.com/dotnet/runtime/commit/c77f524c756bf7c49834d52ce49c235b83b139d6

**Comment by daxian-dbw** [comment]
> > The old style (using IntPtr and void*) forces using _unsafe_ context. Now .Net team introduced `nint` to use _pointer 
>
>There is no reason to avoid unsafe contexts in Pinvokes. In fact, it is an error ([SYSLIB1062](https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1050-1069)) to use [LibraryImportAttribute](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.libraryimportattribute) without AllowUnsafeBlocks set to true.

This I agree. `unsafe` context will be used in the generated partial method anyways.

**Comment by jborean93** [comment]
> so nint should be sufficient, isn't it?

Yep that’s correct, and it being null is the same as IntPtr.Zero here.

**Comment by iSazonov** [comment]
> > > The old style (using IntPtr and void*) forces using _unsafe_ context. Now .Net team introduced `nint` to use _pointer
> > 
> > 
> > There is no reason to avoid unsafe contexts in Pinvokes. In fact, it is an error ([SYSLIB1062](https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1050-1069)) to use [LibraryImportAttribute](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.libraryimportattribute) without AllowUnsafeBlocks set to true.
> 
> This I agree. `unsafe` context will be used in the generated partial method anyways.
> 

To be clear - the PR uses "unsafe" keyword even though it is not necessary. Programming languages exist precisely to avoid unverifiable constructs and `unsafe` is just an indication that the developer is trying to use unverifiable code which can be dangerous. (`unverifiable` term comes from https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code)
From docs https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/unsafe
> The **unsafe** keyword denotes an unsafe context, which is required for any operation involving pointers.

We have no need to do any operations with pointers in the PR.
Although C# does not completely avoid unsafe in some cases no doubt developers of C# languages will strive to do so.

> With regard to the reference dotnet code [dotnet/runtime@c77f524](https://github.com/dotnet/runtime/commit/c77f524c756bf7c49834d52ce49c235b83b139d6), it changes to `WaitForMultipleObjects(..., nint* handles, ...)` because `HANDLE *` is a pointer to another pointer. But for `LPSECURITY_ATTRIBUTES`, it's a pointer to a `SECURITY_ATTRIBUTES` struct (shown below), so `nint` should be sufficient, isn't it?
> 
> ```
> typedef struct _SECURITY_ATTRIBUTES {
>             DWORD nLength;
>             LPVOID lpSecurityDescriptor;
>             BOOL bInheritHandle;
> } SECURITY_ATTRIBUTES, *PSECURITY_ATTRIBUTES, *LPSECURITY_ATTRIBUTES;
> ```

Correct. 
In fact, we can only use blittable types for native function parameters.
This creates two problems - readability and conversion of managed types to blittable types.
So the best thing we can do is to move everything related to pinvokes from the main code to separate files( or a file), as more as possible. 
After migrating a few pinvokes, we can already see that our main code is much more readable, although there are still (readability) issues with the code that directly implements the pinvokes. We need to keep an eye on what the .Net team will offer us in .Net 8.

**Pull request closed without merging**